### ISSUE TITLE
[Bug]: Low Search QPS of ~115, concurrent searches and scaling nodes not working

### ISSUE THREAD

**Issue by ramialt** [milvus-io/milvus#25913]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Environment

```markdown
- Milvus version: 2.2.10
- Deployment mode(standalone or cluster): cluster
- MQ type(rocksmq, pulsar or kafka): pulsar
- SDK version(e.g. pymilvus v2.0.0rc2): 2.2.13
- OS(Ubuntu or CentOS): AL2
- CPU/Memory: 2 AWS m5.4xlarge instances (16 CPU cores, 64 GB memory each)
- GPU: None
- Others:
```

#### Queries are returning much slower than expected, and scaling # of threads/# of nodes does not scale performance. 

For example: Running with 15 threads and searching with 1000 embeddings for each thread, I get all **15k results back in 129 seconds**, which averages out to **~116 queries per second**.
Running with 10 threads and 1500 embeddings per thread gives the same performance, which is odd.

I have scaled to 10 query nodes (to be honest 4 query nodes gives the same performance). When looking in Grafana, it seems like not all available system resources are being used for this search. **Only 2 of the 10 nodes** have segments loaded and **CPU usage maxes out at 11%**, with two query nodes using roughly 5% each. 

#### I would like to be able to use the full system for searches, as I have many vectors to search. 
My understanding is that Kubernetes pods by default use as much CPU and memory as is available in the system, since I have set no limits. Unsure if Milvus has built-in resource limits for this case.

### Expected Behavior

I need to search for about a million vectors, which would take multiple hours at this speed. I need search time to be much faster, and for the additional query nodes I provisioned to be working instead of idle.

Actually, I need to store and search multiple millions of vectors with more complex HNSW index, but search time would be too slow at current QPS. 

### Steps To Reproduce

```markdown
I have 1 million embeddings in this collection and use HNSW index.
Index parameters: {'index_type': 'HNSW', 'metric_type': 'L2', 'params': {'M': 16, 'efConstruction': 100}}

Search parameters:
metric: L2
ef = 100
limit (k) = 2000

Replicas:
Data node - 10
Index node - 10
Query node - 10
(getting same performance as putting above at 4)

Let me know what other information I should provide
```

**Comment by yanliang567** [comment]
There are 1m vectors in the collection, so I think there are only 2-4 segments, so if you load with single replica, only 2-4 querynodes will be loaded with data. I think that's the reason the performance does not turn better after scaling to 10 querynodes. 
My suggestion:
1. try with latest milvus v2.2.12, which fixed/imporved some performance issues. https://github.com/milvus-io/milvus/releases/tag/v2.2.12
2. release the collection, and reload the collection with 2-4 replicas. this helps milvus to balance the search requests in replicas and increase the search QPS.
3. if the performance did not improve as suggested above, please attach some milvus metrics sceenshots, about proxy node, querynode, and runtime.  also attach the birdwatcher backup file would help us understand the segment meta. Check this: https://github.com/milvus-io/birdwatcher for details about how to backup etcd with birdwatcher

/assign @ramialt

**Comment by ramialt** [comment]
Hi, thanks for replica suggestion. Increasing to 20 query nodes and using 10 replicas cut some query time by about half.
(15 threads with 1000 vectors to query each went from 129.55s->68.04s)

I have not yet tried v2.2.12 as this will take some large effort to implement on my end. Are the performance improvements particularly significant? Ideally for my usecase I would still like to scale so that computation time is at least 10-12x less than it is currently, so I have doubts a minor version release has this much boost. I expect there is still more improvements / limits to increase elsewhere, though I will definitely try to update my version.

Particularly, even after scaling to 10 data, 10 index, and 20 query nodes, CPU usage still maxes out at 30%. Maybe search just doesn't use a lot of CPU, but that is not what I would expect.

## Metric screenshots (10 data, 10 index, 20 query pods):
Ran 10 replicas on 1 million embeddings of dim 256.
CPU peak value is when running 30 concurrent queries of 1000 vectors each, which took 126.12s to complete.



### Runtime:
<img width="1291" alt="Screenshot 2023-07-26 at 6 22 40 PM" src="https://github.com/milvus-io/milvus/assets/56455158/f426eabd-16a9-4b0b-8131-0d39c0253de6">

### Proxy:
<img width="1281" alt="Screenshot 2023-07-26 at 6 27 53 PM" src="https://github.com/milvus-io/milvus/assets/56455158/3dc35cc6-8553-42c5-8bce-f73095ea6f28">
<img width="1283" alt="Screenshot 2023-07-26 at 6 28 14 PM" src="https://github.com/milvus-io/milvus/assets/56455158/d6ecebbb-b3c1-4af5-9ae1-0f07bc9a9c39">
<img width="1290" alt="Screenshot 2023-07-26 at 6 28 42 PM" src="https://github.com/milvus-io/milvus/assets/56455158/1ff85682-caec-4aed-9994-d1c1182da647">

### Query Coordinator:
<img width="1290" alt="Screenshot 2023-07-26 at 6 31 17 PM" src="https://github.com/milvus-io/milvus/assets/56455158/44f89eea-52c1-4919-a63d-8c7409cc8f08">
<img width="1281" alt="Screenshot 2023-07-26 at 6 31 41 PM" src="https://github.com/milvus-io/milvus/assets/56455158/54adef63-c428-4628-ae8e-45cda0cdb240">

### Query Node:
<img width="1278" alt="Screenshot 2023-07-26 at 6 29 21 PM" src="https://github.com/milvus-io/milvus/assets/56455158/0426d0f1-67a1-4273-8b67-054844c859c9">
<img width="1267" alt="Screenshot 2023-07-26 at 6 29 58 PM" src="https://github.com/milvus-io/milvus/assets/56455158/68e0b70c-9020-4770-8b0b-06f029221f11">
<img width="1273" alt="Screenshot 2023-07-26 at 6 30 20 PM" src="https://github.com/milvus-io/milvus/assets/56455158/8c99db79-3f0e-45f9-a50a-1ddfc82a565b">
<img width="1285" alt="Screenshot 2023-07-26 at 6 30 31 PM" src="https://github.com/milvus-io/milvus/assets/56455158/76831f7d-5157-44f7-9c73-767ddba20509">
( I am the one releasing the collection from memory at the end there )

### Birdwatcher
I tried to attach backup from birdwatcher but I cannot create backup, I get errors like so:
```
Milvus(by-dev) > backup
Backing up ... 100%(915/914)
backup etcd for prefix  done
http://10.0.126.252:9091/metrics
failed to fetch metrics for datacoord(13), Get "http://10.0.126.252:9091/metrics": dial tcp 10.0.126.252:9091: i/o timeout
http://10.0.97.47:9091/metrics
failed to fetch metrics for datanode(100), Get "http://10.0.97.47:9091/metrics": dial tcp 10.0.97.47:9091: i/o timeout
http://10.0.118.102:9091/metrics
failed to fetch metrics for datanode(101), Get "http://10.0.118.102:9091/metrics": dial tcp 10.0.118.102:9091: i/o timeout
http://10.0.127.188:9091/metrics
failed to fetch metrics for datanode(15), Get "http://10.0.127.188:9091/metrics": dial tcp 10.0.127.188:9091: i/o timeout
http://10.0.168.53:9091/metrics
```
I assume my connection is fine since command `show collections` works fine, so unsure of root cause here.



## Scaling issues
Trying to scale beyond 20 query nodes or 10 data nodes actually worsens performance for some reason.

Using 40 query nodes (and 20 replicas) gives nearly 4x worse performance (and some query nodes die)
Specifically, running the same command as above, 30 concurrent queries of 1000 vectors each, takes 453.27s, with a few threads only taking 124s, and most taking around 450s.

It also resulted in 13 pods being either evicted due to memory pressure or killed with OOMKilled status, which is odd since my memory usage was not close to the system total.

How is it that pods were killed from memory issue in this system, and why is max CPU usage still only at 30%? Attached metrics for this time period below.


## Metric screenshots (10 data, 10 index, 40 query pods):

### Runtime:
<img width="1285" alt="Screenshot 2023-07-26 at 7 01 18 PM" src="https://github.com/milvus-io/milvus/assets/56455158/ba5c627c-daea-4e5d-84ca-034e1eedeffd">
<img width="1288" alt="Screenshot 2023-07-26 at 7 01 38 PM" src="https://github.com/milvus-io/milvus/assets/56455158/5b834357-a7e6-49b6-8e20-2d8c63b572ad">
<img width="1290" alt="Screenshot 2023-07-26 at 7 02 25 PM" src="https://github.com/milvus-io/milvus/assets/56455158/b65fbfb1-c114-41a8-ae7b-bea56d90c8f5">

### Proxy:
<img width="1293" alt="Screenshot 2023-07-26 at 7 02 45 PM" src="https://github.com/milvus-io/milvus/assets/56455158/f8a41447-f748-4460-8cb3-aa587b77c45a">
<img width="1287" alt="Screenshot 2023-07-26 at 7 03 11 PM" src="https://github.com/milvus-io/milvus/assets/56455158/decf7b05-a2f5-43a6-be84-ec5bb984d4ee">
<img width="1295" alt="Screenshot 2023-07-26 at 7 03 24 PM" src="https://github.com/milvus-io/milvus/assets/56455158/0c40bc9d-877c-4ec8-b2fa-bdd8c5aefb19">

### Query Coordinator:
<img width="1298" alt="Screenshot 2023-07-26 at 7 03 48 PM" src="https://github.com/milvus-io/milvus/assets/56455158/051dc8e6-73ce-4e6f-9ccc-b9870a01c482">
<img width="1289" alt="Screenshot 2023-07-26 at 7 03 59 PM" src="https://github.com/milvus-io/milvus/assets/56455158/0ca4d7a2-f840-416a-a46f-6065ad72758f">

### Query Node:
<img width="1295" alt="Screenshot 2023-07-26 at 7 04 19 PM" src="https://github.com/milvus-io/milvus/assets/56455158/a46ab19c-b11e-4115-8774-c61556244f1f">
<img width="1296" alt="Screenshot 2023-07-26 at 7 04 32 PM" src="https://github.com/milvus-io/milvus/assets/56455158/3934699f-e0da-438f-9f76-91c881ab62d0">
<img width="1299" alt="Screenshot 2023-07-26 at 7 04 48 PM" src="https://github.com/milvus-io/milvus/assets/56455158/a3e2310c-ace5-4d43-9a60-e3a94f58ab5e">
<img width="1286" alt="Screenshot 2023-07-26 at 7 05 11 PM" src="https://github.com/milvus-io/milvus/assets/56455158/694db6e9-2a80-42ce-9521-f79b291a3b71">



### My attempted remedy for 40 query pod latency increase: Scaling Data Pods from 10->20
I tried increasing data pods to 20 in case that was the issue, but that just led to even worse performance and errors like below:
```
[search] retry:4, cost: 0.27s, reason: <_InactiveRpcError: StatusCode.UNAVAILABLE, failed to connect to all addresses; last error: UNKNOWN: ipv4:44.227.188.203:80: connection attempt timed out before receiving SETTINGS frame>
[search] retry:10, cost: 60.00s, reason: <_InactiveRpcError: StatusCode.UNAVAILABLE, failed to connect to all addresses; last error: UNAVAILABLE: ipv4:44.227.188.203:80: Socket closed>
```
As well as the Data node "TimeTIck Lag Behind Now (Consumed All)" metric to spike to 6s+ from 250ms. I will not add metric screenshots for this 20 data pod case since this comment is long enough, but I can attach them.



## Summarized questions
* How can I make query nodes use more CPU to be faster?
* How am I hitting memory issues at 88 GiB when I have 128GiB total?
* How can I make this query faster? 
   * My actual usecase uses more complex HNSW index that is about 13x slower to query

P.S. Sorry for the extremely long comment, I wanted to be thorough in log screenshots.

**Comment by yanliang567** [comment]
quick questions for @ramialt :
1. could you please share the milvus yaml file to help us understand the cpu and memory resource limits
2. could you please run kubectl get pod -o wide | grep <your_milvus_release>,  we'd like to check how many pods are sharing the same physical hosts.
3. does milvus run on the exclusive hosts?
please remember to delete the sensitive info if there is any in my request above.

@congqixia any ideas about the birdwatcher backup error above?
/assign @congqixia @liliu-z 
please help to take a look at this issue.

**Comment by xiaofan-luan** [comment]
> ### Is there an existing issue for this?
> * [x]  I have searched the existing issues
> 
> ### Environment
> ```
> - Milvus version: 2.2.10
> - Deployment mode(standalone or cluster): cluster
> - MQ type(rocksmq, pulsar or kafka): pulsar
> - SDK version(e.g. pymilvus v2.0.0rc2): 2.2.13
> - OS(Ubuntu or CentOS): AL2
> - CPU/Memory: 2 AWS m5.4xlarge instances (16 CPU cores, 64 GB memory each)
> - GPU: None
> - Others:
> ```
> 
> ### Current Behavior
> #### Queries are returning much slower than expected, and scaling # of threads/# of nodes does not scale performance.
> For example: Running with 15 threads and searching with 1000 embeddings for each thread, I get all **15k results back in 129 seconds**, which averages out to **~116 queries per second**. Running with 10 threads and 1500 embeddings per thread gives the same performance, which is odd.
> 
> I have scaled to 10 query nodes (to be honest 4 query nodes gives the same performance). When looking in Grafana, it seems like not all available system resources are being used for this search. **Only 2 of the 10 nodes** have segments loaded and **CPU usage maxes out at 11%**, with two query nodes using roughly 5% each.
> 
> #### I would like to be able to use the full system for searches, as I have many vectors to search.
> My understanding is that Kubernetes pods by default use as much CPU and memory as is available in the system, since I have set no limits. Unsure if Milvus has built-in resource limits for this case.
> 
> ### Expected Behavior
> I need to search for about a million vectors, which would take multiple hours at this speed. I need search time to be much faster, and for the additional query nodes I provisioned to be working instead of idle.
> 
> Actually, I need to store and search multiple millions of vectors with more complex HNSW index, but search time would be too slow at current QPS.
> 
> ### Steps To Reproduce
> ```
> I have 1 million embeddings in this collection and use HNSW index.
> Index parameters: {'index_type': 'HNSW', 'metric_type': 'L2', 'params': {'M': 16, 'efConstruction': 100}}
> 
> Search parameters:
> metric: L2
> ef = 100
> limit (k) = 2000
> 
> Replicas:
> Data node - 10
> Index node - 10
> Query node - 10
> (getting same performance as putting above at 4)
> 
> Let me know what other information I should provide
> ```
> 
> ### Milvus Log
> _No response_
> 
> ### Anything else?
> _No response_

Hi ramialt，
Some quick information.
1. you don't need that much datanode and indexnode, put it as 1 should work
2. you only have 2* 16 CPU cores, 64 GB memory physical resources, so change how many pod you have doesn't really helps. All you need is 2 Querynode with 16 request CPUs.
3. to fully utilize your querynode, you just need to create 2 replicas
4. try to change segment size to 2G rather than 512M, that will help on your performance.

please try all these and see if the performance improved

**Comment by ramialt** [comment]
### @yanliang567's response

> could you please share the milvus yaml file to help us understand the cpu and memory resource limits

I used the default milvus.yaml file found [here](https://github.com/milvus-io/milvus/blob/v2.2.10/configs/milvus.yaml). I deployed milvus cluster with helm, maybe that is what you want?
Here are the latest helm values.yaml changes I've tested on:
```
"cluster.enabled": true,
"minio.enabled": false,
"service.port": <not 13590>,
"service.type": "LoadBalancer",
"dataNode.replicas": 10,
"indexNode".replicas": 10,
"queryNode.replicas": 40,
"attu.enabled": true,
"attu.service.port": <different>,
"externalS3.enabled": true,
"externalS3.useIAM": true,
"externalS3.host": 's3.<region>.amazonaws.com',
"externalS3.port": 443,
"externalS3.useSSL": true,
"externalS3.bucketName": <bucketName>
"serviceAccount.create": true,
"serviceAccount.name": <milvus service account name>,
"serviceAccount.annotations": {
        "eks.amazonaws.com/role-arn": <role-arn>
},
"metrics.serviceMonitor.enabled": true
```

> could you please run kubectl get pod -o wide | grep <your_milvus_release>, we'd like to check how many pods are sharing the same physical hosts.

Decided to show output from `kubectl describe nodes` as it shows the list of nodes a bit more compactly, and separates the list of nodes per machine, hope this is okay.
### Machine 1
```
Non-terminated Pods:          (57 in total)
  Namespace                   Name                                   CPU Requests  CPU Limits  Memory Requests  Memory Limits  Age
  ---------                   ----                                   ------------  ----------  ---------------  -------------  ---
  default                     milvus-datacoord-5cc9bccf76-jl7cl      0 (0%)        0 (0%)      0 (0%)           0 (0%)         4d6h
  default                     milvus-datanode-754bc468f6-672zq       0 (0%)        0 (0%)      0 (0%)           0 (0%)         30h
  default                     milvus-datanode-754bc468f6-djqh8       0 (0%)        0 (0%)      0 (0%)           0 (0%)         2d6h
  default                     milvus-datanode-754bc468f6-g42xd       0 (0%)        0 (0%)      0 (0%)           0 (0%)         2d6h
  default                     milvus-datanode-754bc468f6-mbwt7       0 (0%)        0 (0%)      0 (0%)           0 (0%)         2d6h
  default                     milvus-datanode-754bc468f6-r7njt       0 (0%)        0 (0%)      0 (0%)           0 (0%)         4d6h
  default                     milvus-datanode-754bc468f6-r9w86       0 (0%)        0 (0%)      0 (0%)           0 (0%)         30h
  default                     milvus-datanode-754bc468f6-rljz4       0 (0%)        0 (0%)      0 (0%)           0 (0%)         30h
  default                     milvus-datanode-754bc468f6-sfs7g       0 (0%)        0 (0%)      0 (0%)           0 (0%)         30h
  default                     milvus-datanode-754bc468f6-vw49g       0 (0%)        0 (0%)      0 (0%)           0 (0%)         30h
  default                     milvus-etcd-0                          0 (0%)        0 (0%)      0 (0%)           0 (0%)         3d9h
  default                     milvus-etcd-1                          0 (0%)        0 (0%)      0 (0%)           0 (0%)         3d9h
  default                     milvus-etcd-2                          0 (0%)        0 (0%)      0 (0%)           0 (0%)         3d9h
  default                     milvus-indexcoord-79c8b4d5db-xdn82     0 (0%)        0 (0%)      0 (0%)           0 (0%)         4d6h
  default                     milvus-indexnode-d5696ddc5-76z9d       0 (0%)        0 (0%)      0 (0%)           0 (0%)         2d6h
  default                     milvus-indexnode-d5696ddc5-lczkw       0 (0%)        0 (0%)      0 (0%)           0 (0%)         4d6h
  default                     milvus-indexnode-d5696ddc5-s6znt       0 (0%)        0 (0%)      0 (0%)           0 (0%)         2d6h
  default                     milvus-indexnode-d5696ddc5-zggp8       0 (0%)        0 (0%)      0 (0%)           0 (0%)         4d6h
  default                     milvus-proxy-d59b97484-25ql5           0 (0%)        0 (0%)      0 (0%)           0 (0%)         4d6h
  default                     milvus-pulsar-bookie-0                 1 (6%)        0 (0%)      2Gi (3%)         0 (0%)         4d6h
  default                     milvus-pulsar-bookie-1                 1 (6%)        0 (0%)      2Gi (3%)         0 (0%)         4d6h
  default                     milvus-pulsar-bookie-2                 1 (6%)        0 (0%)      2Gi (3%)         0 (0%)         4d6h
  default                     milvus-pulsar-broker-0                 1500m (9%)    0 (0%)      4Gi (6%)         0 (0%)         4d6h
  default                     milvus-pulsar-proxy-0                  1 (6%)        0 (0%)      2Gi (3%)         0 (0%)         4d6h
  default                     milvus-pulsar-recovery-0               1 (6%)        0 (0%)      512Mi (0%)       0 (0%)         4d6h
  default                     milvus-pulsar-zookeeper-0              300m (1%)     0 (0%)      1Gi (1%)         0 (0%)         4d6h
  default                     milvus-querycoord-56d78999b9-72vxz     0 (0%)        0 (0%)      0 (0%)           0 (0%)         4d6h
  default                     milvus-querynode-847894cdb9-6z5bj      0 (0%)        0 (0%)      0 (0%)           0 (0%)         31h
  default                     milvus-querynode-847894cdb9-72jpb      0 (0%)        0 (0%)      0 (0%)           0 (0%)         31h
  default                     milvus-querynode-847894cdb9-7hsg9      0 (0%)        0 (0%)      0 (0%)           0 (0%)         31h
  default                     milvus-querynode-847894cdb9-9c9b2      0 (0%)        0 (0%)      0 (0%)           0 (0%)         4d6h
  default                     milvus-querynode-847894cdb9-brsfx      0 (0%)        0 (0%)      0 (0%)           0 (0%)         31h
  default                     milvus-querynode-847894cdb9-dgnds      0 (0%)        0 (0%)      0 (0%)           0 (0%)         31h
  default                     milvus-querynode-847894cdb9-f9nd5      0 (0%)        0 (0%)      0 (0%)           0 (0%)         31h
  default                     milvus-querynode-847894cdb9-gcrxf      0 (0%)        0 (0%)      0 (0%)           0 (0%)         31h
  default                     milvus-querynode-847894cdb9-k6sps      0 (0%)        0 (0%)      0 (0%)           0 (0%)         32h
  default                     milvus-querynode-847894cdb9-lf7fq      0 (0%)        0 (0%)      0 (0%)           0 (0%)         32h
  default                     milvus-querynode-847894cdb9-n2m7m      0 (0%)        0 (0%)      0 (0%)           0 (0%)         32h
  default                     milvus-querynode-847894cdb9-pnddl      0 (0%)        0 (0%)      0 (0%)           0 (0%)         31h
  default                     milvus-querynode-847894cdb9-pxdmf      0 (0%)        0 (0%)      0 (0%)           0 (0%)         31h
  default                     milvus-querynode-847894cdb9-qpswc      0 (0%)        0 (0%)      0 (0%)           0 (0%)         4d6h
  default                     milvus-querynode-847894cdb9-qt8tf      0 (0%)        0 (0%)      0 (0%)           0 (0%)         32h
  default                     milvus-querynode-847894cdb9-r7tfc      0 (0%)        0 (0%)      0 (0%)           0 (0%)         4d6h
  default                     milvus-querynode-847894cdb9-sr92l      0 (0%)        0 (0%)      0 (0%)           0 (0%)         4d6h
  default                     milvus-querynode-847894cdb9-xvvr2      0 (0%)        0 (0%)      0 (0%)           0 (0%)         31h
  default                     milvus-rootcoord-5588c664b7-ls87k      0 (0%)        0 (0%)      0 (0%)           0 (0%)         4d6h
  kube-system                 aws-node-lmktb                         25m (0%)      0 (0%)      0 (0%)           0 (0%)         4d6h
  kube-system                 coredns-57ff979f67-lpzwk               100m (0%)     0 (0%)      70Mi (0%)        170Mi (0%)     4d6h
  kube-system                 coredns-57ff979f67-qxgvw               100m (0%)     0 (0%)      70Mi (0%)        170Mi (0%)     4d6h
  kube-system                 ebs-csi-controller-7bdf6d99d9-9fddj    60m (0%)      600m (3%)   240Mi (0%)       1536Mi (2%)    4d6h
  kube-system                 ebs-csi-controller-7bdf6d99d9-wfvqd    60m (0%)      600m (3%)   240Mi (0%)       1536Mi (2%)    4d6h
  kube-system                 ebs-csi-node-fnkx6                     30m (0%)      300m (1%)   120Mi (0%)       768Mi (1%)     4d6h
  kube-system                 kube-proxy-x7qct                       100m (0%)     0 (0%)      0 (0%)           0 (0%)         4d6h
  monitoring                  alertmanager-main-1                    14m (0%)      110m (0%)   150Mi (0%)       150Mi (0%)     3d9h
  monitoring                  node-exporter-m79st                    112m (0%)     270m (1%)   200Mi (0%)       220Mi (0%)     3d9h
  monitoring                  prometheus-adapter-54bdfd5865-fhkxk    102m (0%)     250m (1%)   180Mi (0%)       180Mi (0%)     3d9h
  monitoring                  prometheus-k8s-1                       10m (0%)      10m (0%)    450Mi (0%)       50Mi (0%)      3d9h
Allocated resources:
  (Total limits may be over 100 percent, i.e., overcommitted.)
  Resource                    Requests       Limits
  --------                    --------       ------
  cpu                         7513m (47%)    2140m (13%)
  memory                      15544Mi (25%)  4780Mi (7%)
  ephemeral-storage           0 (0%)         0 (0%)
  hugepages-1Gi               0 (0%)         0 (0%)
  hugepages-2Mi               0 (0%)         0 (0%)
  attachable-volumes-aws-ebs  0              0
```
### Machine 2
```
Non-terminated Pods:          (44 in total)
  Namespace                   Name                                   CPU Requests  CPU Limits  Memory Requests  Memory Limits  Age
  ---------                   ----                                   ------------  ----------  ---------------  -------------  ---
  default                     milvus-attu-7f5c45b657-gnp44           0 (0%)        0 (0%)      0 (0%)           0 (0%)         3d7h
  default                     milvus-datanode-754bc468f6-9mxj8       0 (0%)        0 (0%)      0 (0%)           0 (0%)         3d9h
  default                     milvus-indexnode-d5696ddc5-28k2r       0 (0%)        0 (0%)      0 (0%)           0 (0%)         2d6h
  default                     milvus-indexnode-d5696ddc5-7p74t       0 (0%)        0 (0%)      0 (0%)           0 (0%)         2d6h
  default                     milvus-indexnode-d5696ddc5-lc75n       0 (0%)        0 (0%)      0 (0%)           0 (0%)         2d6h
  default                     milvus-indexnode-d5696ddc5-lr8kt       0 (0%)        0 (0%)      0 (0%)           0 (0%)         2d6h
  default                     milvus-indexnode-d5696ddc5-p6bdt       0 (0%)        0 (0%)      0 (0%)           0 (0%)         2d6h
  default                     milvus-indexnode-d5696ddc5-swt2j       0 (0%)        0 (0%)      0 (0%)           0 (0%)         2d6h
  default                     milvus-pulsar-zookeeper-1              300m (1%)     0 (0%)      1Gi (1%)         0 (0%)         4d6h
  default                     milvus-pulsar-zookeeper-2              300m (1%)     0 (0%)      1Gi (1%)         0 (0%)         4d6h
  default                     milvus-querynode-847894cdb9-5c2wj      0 (0%)        0 (0%)      0 (0%)           0 (0%)         29h
  default                     milvus-querynode-847894cdb9-5djf5      0 (0%)        0 (0%)      0 (0%)           0 (0%)         29h
  default                     milvus-querynode-847894cdb9-5h58v      0 (0%)        0 (0%)      0 (0%)           0 (0%)         29h
  default                     milvus-querynode-847894cdb9-6cqkx      0 (0%)        0 (0%)      0 (0%)           0 (0%)         29h
  default                     milvus-querynode-847894cdb9-7qwtx      0 (0%)        0 (0%)      0 (0%)           0 (0%)         29h
  default                     milvus-querynode-847894cdb9-8fzqc      0 (0%)        0 (0%)      0 (0%)           0 (0%)         29h
  default                     milvus-querynode-847894cdb9-d5dbb      0 (0%)        0 (0%)      0 (0%)           0 (0%)         29h
  default                     milvus-querynode-847894cdb9-fctp5      0 (0%)        0 (0%)      0 (0%)           0 (0%)         29h
  default                     milvus-querynode-847894cdb9-fh8fr      0 (0%)        0 (0%)      0 (0%)           0 (0%)         29h
  default                     milvus-querynode-847894cdb9-fzc86      0 (0%)        0 (0%)      0 (0%)           0 (0%)         29h
  default                     milvus-querynode-847894cdb9-j8qmr      0 (0%)        0 (0%)      0 (0%)           0 (0%)         29h
  default                     milvus-querynode-847894cdb9-lzn45      0 (0%)        0 (0%)      0 (0%)           0 (0%)         29h
  default                     milvus-querynode-847894cdb9-m2dg2      0 (0%)        0 (0%)      0 (0%)           0 (0%)         29h
  default                     milvus-querynode-847894cdb9-n2th7      0 (0%)        0 (0%)      0 (0%)           0 (0%)         29h
  default                     milvus-querynode-847894cdb9-n5t87      0 (0%)        0 (0%)      0 (0%)           0 (0%)         2d6h
  default                     milvus-querynode-847894cdb9-pc5xm      0 (0%)        0 (0%)      0 (0%)           0 (0%)         29h
  default                     milvus-querynode-847894cdb9-qghm4      0 (0%)        0 (0%)      0 (0%)           0 (0%)         29h
  default                     milvus-querynode-847894cdb9-r5m7t      0 (0%)        0 (0%)      0 (0%)           0 (0%)         29h
  default                     milvus-querynode-847894cdb9-s7nz6      0 (0%)        0 (0%)      0 (0%)           0 (0%)         29h
  default                     milvus-querynode-847894cdb9-v4sfp      0 (0%)        0 (0%)      0 (0%)           0 (0%)         29h
  default                     milvus-querynode-847894cdb9-x96db      0 (0%)        0 (0%)      0 (0%)           0 (0%)         2d6h
  default                     milvus-querynode-847894cdb9-z2xzc      0 (0%)        0 (0%)      0 (0%)           0 (0%)         29h
  kube-system                 aws-node-6wzrl                         25m (0%)      0 (0%)      0 (0%)           0 (0%)         4d6h
  kube-system                 ebs-csi-node-mmfmc                     30m (0%)      300m (1%)   120Mi (0%)       768Mi (1%)     4d6h
  kube-system                 kube-proxy-mvmwd                       100m (0%)     0 (0%)      0 (0%)           0 (0%)         4d6h
  monitoring                  alertmanager-main-0                    14m (0%)      110m (0%)   150Mi (0%)       150Mi (0%)     3d9h
  monitoring                  alertmanager-main-2                    14m (0%)      110m (0%)   150Mi (0%)       150Mi (0%)     3d9h
  monitoring                  blackbox-exporter-6d998d584d-285ht     30m (0%)      60m (0%)    60Mi (0%)        120Mi (0%)     3d9h
  monitoring                  grafana-78977f78b6-gxn5c               100m (0%)     200m (1%)   100Mi (0%)       200Mi (0%)     3d9h
  monitoring                  kube-state-metrics-84dbfc7b67-gh7cx    40m (0%)      160m (1%)   230Mi (0%)       330Mi (0%)     3d9h
  monitoring                  node-exporter-kpbt5                    112m (0%)     270m (1%)   200Mi (0%)       220Mi (0%)     3d9h
  monitoring                  prometheus-adapter-54bdfd5865-6gmxv    102m (0%)     250m (1%)   180Mi (0%)       180Mi (0%)     3d9h
  monitoring                  prometheus-k8s-0                       10m (0%)      10m (0%)    450Mi (0%)       50Mi (0%)      3d9h
  monitoring                  prometheus-operator-8bc6b485b-8hr6z    110m (0%)     220m (1%)   120Mi (0%)       240Mi (0%)     3d9h
Allocated resources:
  (Total limits may be over 100 percent, i.e., overcommitted.)
  Resource                    Requests     Limits
  --------                    --------     ------
  cpu                         1287m (8%)   1690m (10%)
  memory                      3808Mi (6%)  2408Mi (4%)
  ephemeral-storage           0 (0%)       0 (0%)
  hugepages-1Gi               0 (0%)       0 (0%)
  hugepages-2Mi               0 (0%)       0 (0%)
  attachable-volumes-aws-ebs  0            0
```

> does milvus run on the exclusive hosts?

I'm afraid I do not know what you mean by exclusive hosts. What is this?


### @xiaofan-luan's response

> you don't need that much datanode and indexnode, put it as 1 should work.

Got it. Will probably keep index node higher as I am also testing index build speed, but after this discussion I hope to learn how to scale pods properly.

> you only have 2* 16 CPU cores, 64 GB memory physical resources, so change how many pod you have doesn't really helps. All you need is 2 Querynode with 16 request CPUs.

This confuses me, because adding pods and replicas gave me more performance! When I use less querynodes (my previous data uses #querynode=4, sorry I could not get new data today), doing the same query takes about 130 seconds, compared to 68s with 20 query nodes. 

> with 16 request CPUs

How do I specify this? My understanding with Kubernetes is that if you do not configure limits, then by default pods will use up as much system resources as they need. I have not set any limits.

> to fully utilize your querynode, you just need to create 2 replicas

I am curious, why specifically 2 replicas? 

> try to change segment size to 2G rather than 512M, that will help on your performance

I will try this in my morning and see, sorry I could not today. This will put my current test set data within 1 segment, I am curious as to how this might increase performance? Will post results as soon as I get them

Thank you both for the help, I really appreciate it

**Comment by ramialt** [comment]
### Results
Increasing segment size and reducing to 2 query nodes did give some performance benefit.
Using 15 concurrent requests with 1k vectors each, increasing segment size from 512MB to 2GB **reduced times from 88.6921s -> 71.9313s**.

Unfortunately putting in a resource request for more CPU cores for the query nodes did not help performance. This makes sense since putting no resource request allows the query nodes to use as many resources as needed anyways. Actually, it sometimes decreased performance by about 1.8x longer if both querynodes were put on the same machine.
With the same above settings (using 2GB segment size), it completed in **72.3450 seconds**. 
I set resource requests by setting `queryNode.resources.requests.cpu` in helm values.

All the above was with using **1 data node, 1 index node, and 2 query nodes**.

### Discussion

> All you need is 2 Querynode with 16 request CPUs.

I cannot give query nodes a resource request of all the cpu cores since then no usage would be left for other pods. I was able to get the pods to schedule at 6 cores each, too much above that and they would not be scheduled and be stuck in pending.

Overall, I am still surprised that these searches take so long, especially as I need to search for 1 million vectors, not just these 15k vectors I'm testing with. A colleague of mine used hnswlib directly and managed to create and search 2 million vectors in under 3 hours using similar machines. This was with much higher HNSW index parameters than mine (for comparison, I used their index parameters in Milvus and 5 threads searching just 200 vectors each took >110s in total). I expected less performance than a header-only C++ library, but the difference is very great which is confusing. I am facing similar unexpected runtime issues in index building, which takes many hours.

### Questions:

* After increasing my segment size, I was surprised to find that I was able to query without re-building the index. How does this work, given indexes are supposed to be built per segment?
> does milvus run on the exclusive hosts?

* Still not sure what exclusive hosts means.
* Still curious, why specifically 2 replicas were suggested? Why not scale more pods and more replicas with them? What are the risks?

**Comment by yhmo** [comment]
@ramialt ramialt

Milvus manages data by segments:
- By default, a segment size is around 512MB.
- For each segment, Milvus builds an independent index for it. 
- After you call collection.load(), the segments are loaded into memory of query nodes. Milvus will ensure load balance between query nodes.
- When you call collection.search(), the search engine does ANN search for segments one by one. Search is in-memory.
- For each segment, the search engine generates topk results. Finally, all the topk results are merged into a final topk results and returned to the client.

This is how the Milvus work. It is a distributed system and definitely has system overhead than simply calling interfaces of an HNSW library.

1M vectors with 256-dim, the raw data size is 1M * 256 * 4bytes = 1GB. With larger segment size, you will get fewer segment number. Assume there are 2 query nodes, each segment size is 512MB, then you will get 2 segments, each query node will take 1 segment. But if the segment size is 2GB, then you will get only one segment, only one query node works, the other is idle.
Increasing replica_number can make the idle query node work. Just call collection.load(replica_number=2), 2 copies of the only one segment are loaded into memory, each query node has one copy. This also can increase the QPS of the search. Pre-condition is you have enough RAM to load the copies.


For your questions:
- After increasing my segment size, I was surprised to find that I was able to query without re-building the index. How does this work, given indexes are supposed to be built per segment?
does milvus run on the exclusive hosts?
**After resetting the segment size and restarting the service(so that the configuration takes effect), the segments are not to be changed instantly, they are waiting the compaction action to be triggered (maybe a few minutes). Once a compactions action is triggered, the segments(2 * 512MB) are merged into a large segment(size=1GB). And the index node will begin to build index for the new large segment. Within this time, the search engine is still using the old index of the 512MB segments to search. This is why you can query without rebuilding index. Once the new index is finished, it will be loaded into a query node, and the old index of 512MB will be evicted from the 2 query nodes.**


- Still not sure what exclusive hosts means.
That means the host has no other compute-heavy program running on it.

- Still curious, why specifically 2 replicas were suggested? Why not scale more pods and more replicas with them? What are the risks?
As I mentioned, the replica number is defined according to several factors: segment size, number of query nodes, memory space allocated, etc.
In you use case, if the segment size is changed to 2GB, there is only one segment in the collection, only one query node load the segment. Then if you have 2 query nodes, the second query node will be idle, then we suggest to set the replicas to 2.

**Comment by ramialt** [comment]
Thank you yhmo, xiaofan-luan, yanliang567, and filip-halt (on Slack) for your help, it is much appreciated and I have learned much more about Milvus. I think it is clear now what I was experiencing was not a bug, so I will close this issue. 
I will still be experimenting to see how I can boost the query and index speeds as fast as possible even for these smaller number of vectors, so I may make a discussion post in the future if I have more questions. Thanks so much again
